consensus/src/simplex/actors/voter/round.rs

```diff
@@ -0,0 +1,845 @@
+                // Receiving a certificate for a conflicting proposal means the
+                // leader signed two different payloads for the same (epoch,
+                // view). We immediately flag equivocators, wipe any local vote
+                // accumulation, and rely on the caller to broadcast evidence.
+                let equivocator = self.record_equivocation_and_clear();
+                debug!(
+                    ?equivocator,
+                    ?dropped,
+                    ?retained,
+                    "certificate conflicts with proposal (equivocation detected)"
+                );
+                equivocator
+            }
+            ProposalChange::Skipped => None,
+        }
+    }
+
+    /// Adds a verified notarize vote to the round.
+    ///
+    /// Returns the leader's public key if equivocation is detected (conflicting proposals).
+    pub fn add_verified_notarize(&mut self, notarize: Notarize<S, D>) -> Option<S::PublicKey> {
+        match self.proposal.update(&notarize.proposal, false) {
+            ProposalChange::New | ProposalChange::Unchanged => {}
+            ProposalChange::Replaced { dropped, retained } => {
+                // Once we detect equivocation we clear all votes and return the
+                // leader key so the caller can surface slashable evidence.
+                let equivocator = self.record_equivocation_and_clear();
+                debug!(
+                    ?equivocator,
+                    ?dropped,
+                    ?retained,
+                    "notarize conflicts with certificate proposal (equivocation detected)"
+                );
+                return equivocator;
+            }
+            ProposalChange::Skipped => return None,
+        }
+        self.votes.insert_notarize(notarize);
+        None
+    }
+
+    /// Adds a verified nullify vote to the round.
+    pub fn add_verified_nullify(&mut self, nullify: Nullify<S>) {
+        self.votes.insert_nullify(nullify);
+    }
+
+    /// Adds a verified finalize vote to the round.
+    ///
+    /// Returns the leader's public key if equivocation is detected (conflicting proposals).
+    pub fn add_verified_finalize(&mut self, finalize: Finalize<S, D>) -> Option<S::PublicKey> {
+        match self.proposal.update(&finalize.proposal, false) {
+            ProposalChange::New | ProposalChange::Unchanged => {}
+            ProposalChange::Replaced { dropped, retained } => {
+                let equivocator = self.record_equivocation_and_clear();
+                debug!(
+                    ?equivocator,
+                    ?dropped,
+                    ?retained,
+                    "finalize conflicts with certificate proposal (equivocation detected)"
+                );
+                return equivocator;
+            }
+            ProposalChange::Skipped => return None,
+        }
+        self.votes.insert_finalize(finalize);
+        None
+    }
```

Comment:
I'm confused why a single `finalize` or `notarize` vote would cause you to call `self.proposal.update()`? How can you know that such a vote has the correct proposal?

In my mind `self.proposal.update()` should only be called by `add_verified_notarization()` or `add_verified_finalization()` which both do so via `add_recovered_proposal()`